lib/sgtn-client/api/translation.rb

```diff
@@ -11,42 +11,39 @@ module Core
   end
 
   module Translation
-    def self.getString(component, key, locale)
-      SgtnClient.logger.debug "[Translation.getString]component: #{component}, key: #{key}, locale: #{locale}"
+    def self.getString(component, key, locale, *args, **kwargs)
+      SgtnClient.logger.debug "[Translation.getString]component: #{component}, key: #{key}, locale: #{locale}, args: #{args}"
+      locale = LocaleUtil.get_best_locale(locale)
       str = getTranslation(component, key, locale)
-      if str.nil? && !LocaleUtil.is_source_locale(locale)
-        str = getTranslation(component, key, LocaleUtil.get_source_locale)
+      if str
+        locale = str.locale if str.is_a?(SgtnClient::StringUtil)
+      elsif !LocaleUtil.is_source_locale(locale)
+        locale = LocaleUtil.get_source_locale
+        str = getTranslation(component, key, locale)
+      end
+      return if str.nil?
+
+      if !kwargs.empty?
+        str.localize(locale) % kwargs
+      elsif !args.empty?
+        str.localize(locale) % args
+      else
+        str
       end
-      str
     end
 
     def self.getString_p(component, key, plural_args, locale)
       SgtnClient.logger.debug "[Translation][getString_p]component=#{component}, key=#{key}, locale=#{locale}"
-      str = getTranslation(component, key, locale)
-      if str.nil?
-        unless LocaleUtil.is_source_locale(locale)
-          str = getTranslation(component, key, LocaleUtil.get_source_locale)
-          str.to_plural_s(LocaleUtil.get_source_locale, plural_args) if str
-        end
-      else
-        locale = str.locale if str.is_a?(SgtnClient::StringUtil)
-        str.to_plural_s(locale, plural_args)
-      end
+      getString(component, key, locale, **plural_args)
     end
 
     def self.getString_f(component, key, args, locale, *optionals)
```

Comment:
Is a separate getString_f necessary? Why not call getString directly, passing the intended arguments?

Reply:
This is for compatibility. Can consider to remove them.

Reply:
Same comment as above.